packages/in_app_purchase/android/src/main/java/io/flutter/plugins/inapppurchase/InAppPurchasePlugin.java

```diff
@@ -63,219 +40,56 @@
+    plugin.setupMethodChannel(registrar.activity(), registrar.messenger(), registrar.context());
   }
 
   @Override
-  public void onMethodCall(MethodCall call, Result result) {
-    switch (call.method) {
-      case MethodNames.IS_READY:
-        isReady(result);
-        break;
-      case MethodNames.START_CONNECTION:
-        startConnection((int) call.argument("handle"), result);
-        break;
-      case MethodNames.END_CONNECTION:
-        endConnection(result);
-        break;
-      case MethodNames.QUERY_SKU_DETAILS:
-        querySkuDetailsAsync(
-            (String) call.argument("skuType"), (List<String>) call.argument("skusList"), result);
-        break;
-      case MethodNames.LAUNCH_BILLING_FLOW:
-        launchBillingFlow(
-            (String) call.argument("sku"), (String) call.argument("accountId"), result);
-        break;
-      case MethodNames.QUERY_PURCHASES:
-        queryPurchases((String) call.argument("skuType"), result);
-        break;
-      case MethodNames.QUERY_PURCHASE_HISTORY_ASYNC:
-        queryPurchaseHistoryAsync((String) call.argument("skuType"), result);
-        break;
-      case MethodNames.CONSUME_PURCHASE_ASYNC:
-        consumeAsync((String) call.argument("purchaseToken"), result);
-        break;
-      default:
-        result.notImplemented();
-    }
-  }
-
-  private void startConnection(final int handle, final Result result) {
-    if (billingClient == null) {
-      billingClient = factory.createBillingClient(applicationContext, channel);
-    }
-
-    billingClient.startConnection(
-        new BillingClientStateListener() {
-          private boolean alreadyFinished = false;
-
-          @Override
-          public void onBillingSetupFinished(int responseCode) {
-            if (alreadyFinished) {
-              Log.d(TAG, "Tried to call onBilllingSetupFinished multiple times.");
-              return;
-            }
-            alreadyFinished = true;
-            // Consider the fact that we've finished a success, leave it to the Dart side to validate the responseCode.
-            result.success(responseCode);
-          }
-
-          @Override
-          public void onBillingServiceDisconnected() {
-            final Map<String, Object> arguments = new HashMap<>();
-            arguments.put("handle", handle);
-            channel.invokeMethod(MethodNames.ON_DISCONNECT, arguments);
-          }
-        });
-  }
-
-  private void endConnection(final Result result) {
-    if (billingClient != null) {
-      billingClient.endConnection();
-      billingClient = null;
-    }
-    result.success(null);
-  }
-
-  private void isReady(Result result) {
-    if (billingClientError(result)) {
-      return;
-    }
-
-    result.success(billingClient.isReady());
+  public void onAttachedToEngine(FlutterPlugin.FlutterPluginBinding binding) {
+    setupMethodChannel(
+        null, binding.getFlutterEngine().getDartExecutor(), binding.getApplicationContext());
```

Comment:
(subjective, optional): It's a little easier to read if the primitive params are annotated. This is totally an optional/subjective suggestion though, feel free to ignore.

```suggestion
        /*activity=*/null, binding.getFlutterEngine().getDartExecutor(), binding.getApplicationContext());
```